### DISCUSSION TITLE
Wrong alignment when pasting YAML into Asciidoc

### DISCUSSION THREAD

**Discussion by mramendi** [pulsar-edit/.github#274] [Core Application & Packages]
Hello,

So, I have Pulsar, currently 1.116.0. I use it to write documentation in Asciidoc and I need to unclude YAML examples. So, I have the scaffolding:

```
[source,yaml,subs="attributes+"]
----

----

```
I try to paste a YAML block into the space. But the YAML block is aligned two spaces to the right, and I have to un-align it using Control+[ .

How can I make the editor paste YAML at the level that the cursor is at?

**Comment by savetheclocktower** [comment]
**Editor: Paste Without Reformatting** is the command you want; on macOS it's bound to <kbd>Cmd-Shift-V</kbd>, and I imagine it's got a similar keybinding on other platforms.

**Comment by mramendi** [comment]
I don't really want "without" reformatting as much as "not assume an offset but paste from where I am now, reformatting accordingly" - is that possible

**Reply by savetheclocktower** [reply]
I started to write another answer, got curious, and ended up diving into this for about a half-hour.

You'll notice that, if you change the `yaml` in your code example to something else before pasting, the text doesn't indent when you paste.

TextMate-style grammars (like the one from `language-asciidoc`) have a setting called `editor.increaseIndentPattern` that governs when a given line should be indented relative to the previous line. I was confused because this setting is not specified for AsciiDoc. It took me a while to realize that, because the code in your sample creates a YAML “injection” in between the two `----` lines, the pasted text is also assessed according to the rules of the YAML grammar.

The YAML grammar, when asked “hey, should I indent by one level if the previous line's text was `----`?”, will answer _yes_. Never mind the fact that `----` in your file isn't YAML and shouldn't be considered in this case.

So this is a bug in our legacy TextMate grammar system. I'm not sure if this could be fixed without lots of disruption, but I can at least offer you a workaround that will redefine YAML's indentation rule to avoid this scenario.

[Open your init file](https://docs.pulsar-edit.dev/customizing-pulsar/the-init-file/) and paste this into it:

```js
atom.config.set(
  'editor.increaseIndentPattern',
  '^(?!----)\\s*.*(:|-) ?(&\\w+)?(\\{[^}"\']*|\\([^)"\']*)?$',
  { scopeSelector: '.source.embedded.yaml' }
)
```

Then save the file and relaunch Pulsar (or run the **Window: Reload** command).

That should preserve the ordinary indentation rules inside of embedded YAML, _except_ for cases where the preceding line starts with `----`. It seems to work for me locally.

Our newer Tree-sitter grammar system doesn't have this issue, and if there's ever [a mature Tree-sitter parser for AsciiDoc](https://github.com/cpkio/tree-sitter-asciidoc), perhaps someone can make a successor grammar.
